src/app/views/query-runner/query-input/AutoComplete.tsx

```diff
@@ -0,0 +1,344 @@
+        break;
+
+      default:
+        break;
+    }
+  }
+
+  public onKeyDown = (e: any) => {
+    const { activeSuggestion, filteredSuggestions, showSuggestions } = this.state;
+
+    switch (e.keyCode) {
+      case KeyCodes.enter:
+        if (showSuggestions) {
+          const selected = filteredSuggestions[activeSuggestion];
+          this.appendSuggestionToUrl(selected);
+        } else {
+          this.props.runQuery();
+        }
+        break;
+
+      case KeyCodes.up:
+        if (activeSuggestion === 0) {
+          return;
+        }
+        this.setState({ activeSuggestion: activeSuggestion - 1 });
+        break;
+
+      case KeyCodes.down:
+        if (activeSuggestion - 1 === filteredSuggestions.length) {
+          return;
+        }
+        this.setState({ activeSuggestion: activeSuggestion + 1 });
+        break;
+
+      default:
+        break;
+    }
+
+  };
+
+  public displayLinkOptions = () => {
+    const { autoCompleteOptions, sampleQuery: { selectedVerb } } = this.props;
+    if (autoCompleteOptions) {
+      const suggestions: string[] = [];
+      const parameters = autoCompleteOptions.parameters;
+      if (parameters) {
+        const parametersWithVerb = parameters.find(parameter => parameter.verb === selectedVerb.toLowerCase());
+        if (parametersWithVerb) {
+          parametersWithVerb.links.forEach((value: string) => {
+            suggestions.push(value);
+          });
+
+          this.setState({
+            filteredSuggestions: suggestions,
+            suggestions,
+            showSuggestions: true,
+            compare: ''
+          });
+        }
+      }
+    }
+  }
+
+  public getQueryParameters = () => {
+    const { autoCompleteOptions, sampleQuery: { selectedVerb } } = this.props;
+    if (autoCompleteOptions) {
+      const suggestions: string[] = [];
+      const parameters = autoCompleteOptions.parameters;
+      if (parameters) {
+        const parametersWithVerb = parameters.find(parameter => parameter.verb === selectedVerb.toLowerCase());
+        if (parametersWithVerb) {
+          parametersWithVerb.values.forEach((value: any) => {
+            suggestions.push(value.name);
+          });
+
+          this.setState({
+            filteredSuggestions: suggestions,
+            suggestions,
+            showSuggestions: true,
+            compare: ''
+          });
+        }
+      }
+    }
+  }
+
+  public componentDidUpdate = (prevProps: IAutoCompleteProps) => {
+    if (prevProps.autoCompleteOptions !== this.props.autoCompleteOptions) {
+      if (this.props.autoCompleteOptions) {
+        this.performLocalSearch(this.props.sampleQuery.sampleUrl);
+      }
+    }
+  }
+
+  private getParameterEnums = (url: string) => {
+    const { autoCompleteOptions, sampleQuery: { selectedVerb } } = this.props;
+    if (!autoCompleteOptions) {
+      return;
+    }
+    const parameters = autoCompleteOptions.parameters;
+    const param = url.split('$').pop()!.split('=')[0];
```

Comment:
This line should go into the `if (parameters)` block below where it is used, so that we avoid running the relatively expensive string operations if `parameters` is not valid.